cpp/src/arrow/filesystem/test_util.cc

```diff
@@ -453,9 +453,11 @@ void GenericFileSystemTest::TestMoveDir(FileSystem* fs) {
   AssertAllFiles(fs, {"EF/ghi", "KL/CD/def", "KL/abc"});
 
   // Destination is a non-empty directory
-  ASSERT_RAISES(IOError, fs->Move("KL", "EF"));
-  AssertAllDirs(fs, {"EF", "KL", "KL/CD"});
-  AssertAllFiles(fs, {"EF/ghi", "KL/CD/def", "KL/abc"});
+  if (!allow_move_dir_over_non_empty_dir()) {
+    ASSERT_RAISES(IOError, fs->Move("KL", "EF"));
+    AssertAllDirs(fs, {"EF", "KL", "KL/CD"});
+    AssertAllFiles(fs, {"EF/ghi", "KL/CD/def", "KL/abc"});
+  }
```

Comment:
Is it worth testing the other side of this? (Presumably it'd succeed and we'd have `EF/KL/{abc, CD/def}`?)

Reply:
Yes, will do.